### DISCUSSION TITLE
Authentication errors when upgrading from v8.0.0-R24031302 to v8.0.0-R24032503 / v8.0.1

### DISCUSSION THREAD

**Discussion by JacoJordaan** [MarimerLLC/csla#3815] [Questions]
I have a pure server-side Blazor up that was successfully upgraded from .net 7 using csla 7 to .net 8 using v8.0.0-R24031302.
Since I am only using Blazor server-side, I have made no configuration changes for the new Blazor render modes.

Upgrading from v8.0.0-R24031302 to v8.0.0-R24032503 or v8.0.0 or v8.0.1 returns 2 exceptions. I am not sure if it is related.

The first one is due to the ApplicationContext.User.Identity that returns null. Previously this always had a value. 
Screenshot of issue where the ApplicationContext was injected into a service:
![image](https://github.com/MarimerLLC/csla/assets/43467554/a750b212-e315-4f63-b5ba-52a1fae13603)

The next one gives me the following error:
````
System.InvalidOperationException: Do not call GetAuthenticationStateAsync outside of the DI scope for a Razor component. Typically, this means you can call it only within a Razor component or inside another DI service that is resolved for a Razor component.
   at Csla.DataPortal`1.Create(Type objectType, Object criteria)
   at Csla.DataPortal`1.Create(Object[] criteria)

````

Is there any configuration changes required to use the latest version of csla or is this a possible bug?

**Comment by kcabral817** [comment]
I did the same thing but rolled back because of the error, so I'm interested in this too.

**Comment by StefanOssendorf** [comment]
Where does the service live? Is it in a hosted service besides the blazor stack?
Is it possible for you to create a minimal working example which provokes that exact error to make investigation easier?

**Reply by JacoJordaan** [reply]
I have created a minimal working example attached.
[BlazorServerTestCsla8.zip](https://github.com/MarimerLLC/csla/files/15023657/BlazorServerTestCsla8.zip)

I created this initialy in .net7 with csla 7.0.4 and everything was working.
Upgraded to .net8 with csla 8.0.1 and configuration changes for csla 8.

I am not able to replicate the ApplicationContext.User.Identity == null issue, but I still receive the error
````
Do not call GetAuthenticationStateAsync outside of the DI scope for a Razor component. Typically, this means you can call it only within a Razor component or inside another DI service that is resolved for a Razor component.
````

**Reply by rockfordlhotka** [reply]
fwiw, @mtavares628 ran into this and did find a workaround - something to do with how the custom authentication state manager is implemented iirc.

**Reply by mtavares628** [reply]
@rockfordlhotka If this is related to my [issue](https://github.com/MarimerLLC/csla/discussions/3749#discussioncomment-8909737)  where the identity is null, I never actually found a solution (Note that the title for that discussion doesn't match where the problem was truly found). I got pulled into another project for the last few weeks, so I haven't been able to follow up with you about it.  But the issue was introduced for me in pull request  #3632 and in particular this commit from it: https://github.com/MarimerLLC/csla/pull/3632/commits/d3aa7cc74e8156a534bf786c1b57c0e33d111c8d.

My code still ran after this but the performance was terrible, and I think it's because it's throwing an exception from calling GetAuthenticationState before SetAuthenticationState which may be related to this problem.  I'm hoping to be finished with my other project by next week, and can revisit this to see what's going on and how we can fix it.

**Comment by rockfordlhotka** [comment]
> Is there any configuration changes required to use the latest version of csla or is this a possible bug?

Yes, there's a config change that's necessary. CSLA 8 defaults to using the new state management subsystem, and if you are doing a _pure_ render mode (like .NET 7) app you need to use the "in memory" state management.

```c#
builder.Services.AddCsla(o => o
  .AddAspNetCore()
  .AddServerSideBlazor(ssb => ssb.UseInMemoryApplicationContextManager = true));
```

cc @russblair as this is what we were just discussing

**Reply by russblair** [reply]
To recreate this issue, I upgraded the BlazorExample.sln from CSLA 8.0.0-R24021201 to CSLA 8.1.0.

I had to change the CslaStateController.cs to override the new StateManager.Get and I had to remove the Home.razor code related to StateManager.IsStateAvailable.

When I run it and click on the “List people” menu option it produces the error “Do not call GetAuthenticationStateAsync outside of the DI scope for a Razor component. Typically, this means you can call it only within a Razor component or inside another DI service that is resolved for a Razor component.”

**Reply by russblair** [reply]
The error is the same after changing the BlazorExample/Program.cs to .AddServerSideBlazor(ssb => ssb.UseInMemoryApplicationContextManager = true)

**Reply by rockfordlhotka** [reply]
Thanks for doing the research @russblair - I'll look into this further.

**Reply by swegele** [reply]
Just bumped into this same error after upgrading my all-server-side-interactive blazor project.  

```csharp
builder.Services.AddCsla(o => o
  .AddAspNetCore()
  .AddServerSideBlazor(bsco => bsco
    .UseInMemoryApplicationContextManager = true)
  .Security(so => so.FlowSecurityPrincipalFromClient = true));
```

**Comment by kcabral817** [comment]
IMHO, I think .NET 8 release is way more than just a breaking change.  I get the sense from reading the conversations here that it is a fundamental shift for Blazor communication.  Does anyone know why such a change is warranted and what benefit it will bring?

**Reply by StefanOssendorf** [reply]
As far as I understand it's necessary to support the new rendering modes.

**Reply by rockfordlhotka** [reply]
The new render modes in Blazor 8 are an amazing feature - once they are understood.

In my experience, most apps have lots of read-only pages that show data, graphs, charts, and all sorts of stuff. There's no reason to set up SignalR pipelines or download wasm runtimes to display data - the web was great at doing this from day 1 back in the 1990s.

Blazor _was_ requiring that _all_ pages have either SignalR or wasm, which is total overkill for all those pages. 

The new render model allows us to create server-static pages that don't have any of that overhead, just sending html/css to the browser. And for longer data reports, you can use streaming, where the user sees the data as it is sent to the browser over time. Very good experience imo.

For pages that _do_ need interactivity, you can use server-interactive or wasm-interactive. The thing is, both of those have some upfront setup time - either to create the SignalR pipeline or to download the wasm files. To give the user a more immediate feel, you can use InteractiveAuto.

InteractiveAuto does this (on the very first use of an interactive page):

1. Render using server-static to show the user something _immediately_
2. Render using server-interactive, because setting up a SignalR pipeline is fairly fast; and behind the scenes while the user is using the page, the wasm files are being downloaded
3. On a subsequent visit to the page by the user, render using server-static to show the user something _immediately_; the wasm files are being JIT compiled behind the scenes
4. On that subsequent visit, render the page using wasm-interactive to shift the workload to the client device and entirely off the server
5. On _future_ visits to the page, render the page using wasm-interactive because it has been downloaded and compiled, so it can be just launched directly

As long as you have a reliable way to [detect the current render mode](https://blog.lhotka.net/2024/03/30/Blazor-8-Render-Mode-Detection) and understand this workflow, you can make a very compelling user experience that leverages interactivity.

For example:

1. On server-static rendering, render placeholders so the user knows something is coming
2. On interactive rendering, render the data and show all the buttons and other stuff the user can see and use

This avoids double-loading the data (on server-static and then again in interactive mode) by only loading the data in interactive mode, but giving the user a decent "loading" visual in the meantime.

**Comment by swegele** [comment]
I'm wondering if .NET 8 changed the error and message such that this little test we added in CSLA 6 and .NET 6 is no longer catching the condition (which would explain why it happens for both inmemory and not) :

See both:  
- `Csla.AspNetCore\Blazor\ApplicationContextManagerInMemory`
- `Csla.AspNetCore\Blazor\ApplicationContextManagerBlazor`


```csharp

private void InitializeUser()
{
  Task<AuthenticationState> task = default;
  try
  {
    task = AuthenticationStateProvider.GetAuthenticationStateAsync();
  }
  catch (InvalidOperationException ex)
  {
    task = Task.FromResult(new AuthenticationState(UnauthenticatedPrincipal));
    string message = ex.Message;
    if (message.Contains(nameof(AuthenticationStateProvider.GetAuthenticationStateAsync))
        && message.Contains(nameof(IHostEnvironmentAuthenticationStateProvider.SetAuthenticationState)))
    {
      SetHostPrincipal(task);
    }
    else
    {
      throw;
    }
  }
  AuthenticationStateProvider_AuthenticationStateChanged(task);
}

```

If I get some time I will debug to see if we are hitting that condition and passing over it because the error message possibly no longer mentions the "SetAuthenticationState" part like it used to?

**Reply by swegele** [reply]
Gotta love open source :-)   See the [aspnet core code here](https://github.com/dotnet/aspnetcore/blob/87e324a61dcd15db4086b8a8ca7bd74ca1e0a513/src/Components/Server/src/Circuits/ServerAuthenticationStateProvider.cs#L16).
This explanation is looking promising:



```csharp
public class ServerAuthenticationStateProvider : AuthenticationStateProvider, IHostEnvironmentAuthenticationStateProvider
{
    private Task<AuthenticationState> _authenticationStateTask;

    /// <inheritdoc />
    public override Task<AuthenticationState> GetAuthenticationStateAsync()
        => _authenticationStateTask
        ?? throw new InvalidOperationException($"Do not call {nameof(GetAuthenticationStateAsync)} outside of the DI scope for a Razor component. Typically, this means you can call it only within a Razor component or inside another DI service that is resolved for a Razor component.");

```

**Reply by swegele** [reply]
Yep - click on "Blame" and you see [this...viola](https://github.com/dotnet/aspnetcore/commit/392a0902fe2293fd8ba23976c2c594f47fb05e2c) 

Mr. Halter did a helpful thing and "improved the exception message" :-)

![image](https://github.com/MarimerLLC/csla/assets/6868671/c9292589-9c24-4bea-b9bd-039a77027915)

**Reply by StefanOssendorf** [reply]
The catch quoted by @swegele can be removed. It's _never_ hit because the task is not awaited. So there won't be any exceptions. That's probably a bug.

**Reply by swegele** [reply]
I remembered this because we were hitting it when back in .NET 6 and CSLA 6.  It was only with the ServerAuthenticationStateProvider.
Not sure about your point on not awaiting but we certainly were hitting this and that code fixed it.

**Reply by mtavares628** [reply]
@swegele I can confirm with @StefanOssendorf that the catch is never hit because it's not awaited. This was part of the issue that I outlined in [3749](https://github.com/MarimerLLC/csla/discussions/3749#discussioncomment-8909737).  The issue was introduced for me in pull request https://github.com/MarimerLLC/csla/pull/3632 and in particular this commit from it: https://github.com/MarimerLLC/csla/commit/d3aa7cc74e8156a534bf786c1b57c0e33d111c8d where it is making LocalProxy threadsafe.  After this change I started getting the InvalidOperationException that InitializeUser is supposed to catch, but it never does because it's not awaited.

**Reply by swegele** [reply]
@StefanOssendorf  if you are curious you can see the discussion back in Feb 2022 that caused those lines...we definitely were hitting that error but I think @TheCakeMonster was confused as well WHY we were hitting it.  I did a simple demo and uploaded it for review showing it was being hit.

https://github.com/MarimerLLC/csla/pull/2739/files/cae93bec07ba88209b7ea3e44c161c879cfd3f0e#r796945895

**Reply by swegele** [reply]
@mtavares628  @StefanOssendorf  - OK well maybe it was an issue at one point?  I'll trust you guys that it is not relevant anymore.

**Reply by mtavares628** [reply]
@swegele, I think it's still relevant, I think we are saying that the code isn't being hit, but it should be.  It's a bug in that the call needs to be async in order for the exception to be caught.

**Reply by swegele** [reply]
OK, if it was properly caught then would we still need to change the code to check for the correct error message now that it is changed?

P.S.  Seems a bit brittle/fragile to have code testing the .NET error message text, but I don't know a better way.

**Reply by StefanOssendorf** [reply]
There are cases when the catch can be hit. That's the case when `GetAuthenticationStateAsync()` like in your other quote is _not_ async. In that case the exceptions is thrown immediately without an await. In _all other_ cases - with a proper async implementation - the catch is never hit because the exception is waiting in the task to be awaited.
I'll create a bug/research issue to keep track of that.

**Reply by swegele** [reply]
I was able to get rid of the error by changing the InitializeUser method to the following:

```csharp
    private async Task InitializeUser()
    {
      Task<AuthenticationState> task = default;
      try
      {
        await (task = AuthenticationStateProvider.GetAuthenticationStateAsync());
      }
      catch (InvalidOperationException ex)
      {
        task = Task.FromResult(new AuthenticationState(UnauthenticatedPrincipal));
        string message = ex.Message;
        if (message.Contains(nameof(AuthenticationStateProvider.GetAuthenticationStateAsync) + " outside of the DI scope for a Razor component"))
        {
          SetHostPrincipal(task);
        }
        else
        {
          throw;
        }
      }
      AuthenticationStateProvider_AuthenticationStateChanged(task);
    }

```

**Reply by swegele** [reply]
@russblair  @rockfordlhotka  -  Hey all, I followed Russ' breadcrumb path

- Upgrading the BlazorExample project to 8.1 nuget
- Changed the CslaStateController line to this:  `public override StateResult Get(long lasttouched) => base.Get(lasttouched);`
- Commented out this on Home: `//IsStateReady = StateManager.IsStateAvailable;`
Note - I did not change Program.cs at all.  

Run project and go to list page - 
BOOM see the error

Next 
- Make the change in the message just above this post to these files `InitializeUser` method: 
  - Csla.AspNetCore\Blazor\ApplicationContextManagerInMemory
  - Csla.AspNetCore\Blazor\ApplicationContextManagerBlazor
- Paste this into the BlazorExample project properties : Build : Events : Post-build event
```
xcopy /y /d  "C:\repos\csla\Bin\Debug\net8.0\Csla.dll" "$(TargetDir)"
xcopy /y /d  "C:\repos\csla\Bin\Debug\net8.0\Csla.pdb" "$(TargetDir)"
xcopy /y /d  "C:\repos\csla\Bin\Debug\net8.0\Csla.AspNetCore.dll" "$(TargetDir)"
xcopy /y /d  "C:\repos\csla\Bin\Debug\net8.0\Csla.AspNetCore.pdb" "$(TargetDir)"
xcopy /y /d  "C:\repos\csla\Bin\Debug\net8.0\Csla.Blazor.dll" "$(TargetDir)"
xcopy /y /d  "C:\repos\csla\Bin\Debug\net8.0\Csla.Blazor.pdb" "$(TargetDir)"
```
This copied the dlls you just compiled over the nuget ones at after build of blazor project so you get new behavior.

Now run and the blazor list page loads without error.

**Reply by rockfordlhotka** [reply]
@swegele if you make `InitializeUser` return `async Task`, then you can't call it from the constructor. How did you get around that?

**Reply by swegele** [reply]
I _think_ it simply gets treated as normal sync...so it seems a little confusing to me to mark the `InitializeUser` method as async since the only caller to it is the sync CTOR.  But maybe it does seem more correct to have the method be async if it ever needed to be called from elsewhere.

@StefanOssendorf  had the nice suggestion of adding this `_ = InitializeUser();` to make the compiler message go away in the CTOR :-)

Anyway, for my situation the issue was the changed error message from microsoft not the task await thing.

**Comment by rockfordlhotka** [comment]
Will be fixed in #3903 (CSLA 9)

**Reply by rockfordlhotka** [reply]
Oops, not fixed 😳

**Comment by rockfordlhotka** [comment]
Should be fixed in #3964
